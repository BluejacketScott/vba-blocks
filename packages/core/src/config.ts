import { join } from 'path';
import { homedir } from 'os';
import { PackageInfo } from './packages';

/**
 * Keep track of absolute paths to various resources and package resolution
 * 
 * Currently hardcoded, but I imagine an .rc or .config.js approach in the future
 */

export interface Config {
  /**
   * cwd for processing
   */
  cwd: string;

  /**
   * absolute path to build directory
   * (build contains binary output of build process)
   */
  build: string;

  /**
   * absolute path to scripts directory
   * (contains applescript and vbs)
   */
  scripts: string;

  /**
   * absolute path to addins directory
   */
  addins: string;

  /**
   * absolute path to cache directory
   * (default is .vba-blocks in home directory)
   */
  cache: string;

  /**
   * absolute path to local registry
   * and url for git remote
   */
  registry: { local: string; remote: string };

  /**
   * resolve package from remote source
   * (default is packages.vba-blocks.com/...)
   */
  resolveRemotePackage: (pkg: PackageInfo) => string;

  /**
   * resolve package locally
   * (default is in {cache}/packages/...)
   */
  resolveLocalPackage: (pkg: PackageInfo) => string;

  /**
   * resolve expanded package ("source")
   * (default is in {cache}/sources/...)
   */
  resolveSource: (pkg: PackageInfo) => string;
}

export async function loadConfig(): Promise<Config> {
  const cwd = process.cwd();
  const build = join(cwd, 'build');
  const scripts = join(__dirname, '../scripts');
  const addins = join(__dirname, '../../addin/build');
  const cache = join(homedir(), '.vba-blocks');
  const registry = {
    local: join(cache, 'registry'),
    remote: 'https://github.com/vba-blocks/registry.git'
  };

  const resolveRemotePackage = pkg =>
    `https://packages.vba-blocks.com/${pkg.name}/v${pkg.version}.tar.gz`;
  const resolveLocalPackage = pkg =>
    join(cache, `packages/${pkg.name}/v${pkg.version}.tar.gz`);
  const resolveSource = pkg =>
    join(cache, `sources/${pkg.name}/v${pkg.version}`);

  const config: Config = {
    cwd,
    build,
    scripts,
    addins,
    cache,
    registry,
    resolveRemotePackage,
    resolveLocalPackage,
    resolveSource
  };

  return config;
}
