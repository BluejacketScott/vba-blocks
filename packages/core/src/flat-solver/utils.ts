export type Iterable = any;

export function forEach(
  iterator: Iterable,
  fn: (item: any, index: number) => any
) {
  let index = 0;
  for (const item of iterator) {
    fn(item, index++);
  }
}

export function map(
  iterator: Iterable,
  fn: (item: any, index: number) => any
): any[] {
  let index = 0;
  const results = [];
  for (const item of iterator) {
    results.push(fn(item, index++));
  }
  return results;
}

export function pluck(values: Iterable, key: string): any[] {
  return map(values, value => value && value[key]);
}

export function reduce<T>(
  iterator: Iterable,
  fn: (memo: T, item: any) => T,
  memo: T
) {
  for (const item of iterator) {
    memo = fn(memo, item);
  }

  return memo;
}

export function unique(values: Iterable): any[] {
  const asSet = new Set(values);
  return Array.from(asSet);
}
