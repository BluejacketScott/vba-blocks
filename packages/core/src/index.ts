export {
  Version,
  Dependency,
  Feature,
  Reference,
  Source,
  Target,
  Manifest
} from './manifest';
export { Config } from './config';
