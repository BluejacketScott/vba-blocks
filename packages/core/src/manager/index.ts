import { Config } from '../config';
import { Dependency } from '../manifest';
import {
  isRegistryDependency,
  isPathDependency,
  isGitDependency
} from '../manifest/dependency';
import { Registration } from './registration';
import * as registry from './registry-manager';
import * as path from './path-manager';
import * as git from './git-manager';

export { Registration, registry, path, git };
export { getRegistrationId } from './registration';

export default class Manager {
  config: Config;

  constructor(config: Config) {
    this.config = config;
  }

  async update() {
    await registry.update(this.config);
  }

  async resolve(dependency: Dependency): Promise<Registration[]> {
    if (isRegistryDependency(dependency)) {
      return registry.resolve(this.config, dependency);
    } else if (isPathDependency(dependency)) {
      return path.resolve(this.config, dependency);
    } else {
      return git.resolve(this.config, dependency);
    }
  }

  async fetch(registration: Registration) {
    // TODO Encode enough info into registration to determine registry, path, or git
  }
}
