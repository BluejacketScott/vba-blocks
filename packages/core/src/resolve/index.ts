import { Config } from '../config';
import { Manifest } from '../manifest';
import { loadLockfile, writeLockfile } from './lockfile';

export interface DependencyGraph {}

export default async function resolve(
  config: Config,
  manifest: Manifest
): Promise<DependencyGraph> {
  return {};
}
