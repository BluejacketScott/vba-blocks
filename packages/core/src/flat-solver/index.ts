import * as semver from 'semver';
import * as Logic from 'logic-solver';
import { forEach, map, pluck } from './utils';
import {
  resolveDependencies,
  FlatGraph,
  FlatGraphNode
} from './resolve-dependencies';

export { resolveDependencies, FlatGraph, FlatGraphNode };

export type Version = string;
export type Dependency = { name: string; version: Version };

export interface Resolver {
  (name: string): Promise<PackageInfo[]>;
}

export interface PackageInfo {
  id?: string;
  name: string;
  version: Version;
  dependencies: Dependency[];
}

export interface Solution {
  success: boolean;
  result?: string[];
}

/**
 * Solve for the resolved flat dependency tree given a set of root dependencies
 */
export async function solve(
  dependencies: Dependency[] | { [name: string]: Version },
  resolver: Resolver
): Promise<Solution> {
  if (!Array.isArray(dependencies))
    dependencies = toArray<Dependency>(dependencies);

  const graph = await resolveDependencies(dependencies, resolver);
  const solver = new Logic.Solver();

  forEach(graph.values(), node => {
    const { name, versions, required } = node;
    const ids = pluck(versions, 'id');

    if (required) {
      solver.require(Logic.exactlyOne(...ids));
    } else {
      solver.require(Logic.atMostOne(...ids));
    }
  });

  const reversed = [...graph.values()].reverse();
  forEach(reversed, node => {
    const { versions } = node;

    forEach(versions, info => {
      const { id } = info;
      forEach(info.dependencies, dependency => {
        const matching = getMatching(dependency, graph);
        solver.require(Logic.implies(id, Logic.or(...matching)));
      });
    });
  });

  const solution = solver.solve();
  if (!solution) return { success: false };

  return {
    success: true,
    result: solution.getTrueVars()
  };
}

function getMatching(dependency, graph) {
  const { name, version: req } = dependency;
  const versions = graph.get(name).versions;

  const matching = versions
    .filter(({ version }) => semver.satisfies(version, req))
    .map(version => version.id);

  return matching;
}

function toArray<T>(dependencies): T[] {
  return map(Object.entries(dependencies), ([name, version]) => {
    return { name, version };
  });
}
