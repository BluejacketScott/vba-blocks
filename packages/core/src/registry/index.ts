import { join, dirname, basename } from 'path';
import * as readline from 'readline';
import { createReadStream, ensureDir, exists } from 'fs-extra';
import { clone, pull } from '../utils/git';
import { Config } from '../config';
import { Version, Dependency } from '../manifest';

export interface Registration {
  name: string;
  version: Version;
  dependencies: Dependency[];
  features: { [name: string]: string[] };
  checksum?: string;
  yanked?: boolean;
}

export async function getRegistered(
  config: Config,
  name: string
): Promise<Registration[]> {
  return new Promise<Registration[]>((resolve, reject) => {
    const path = getPath(config, name);
    const registrations: Registration[] = [];

    const reader = readline.createInterface({
      input: createReadStream(path)
    });
    reader.on('line', line => {
      const { name, vers, deps, cksum, features, yanked } = JSON.parse(line);
      const dependencies = deps.map(dependency => {
        const { name, req, features, optional, default_features } = dependency;
        return { name, range: req, features, optional, default_features };
      });

      registrations.push({
        name,
        version: vers,
        dependencies,
        features,
        checksum: cksum,
        yanked
      });
    });
    reader.on('close', () => resolve(registrations));
    reader.on('error', reject);
  });
}

export async function updateRegistry(config: Config): Promise<void> {
  const { local, remote } = config.registry;

  if (!await exists(local)) {
    const dir = dirname(local);
    await ensureDir(dir);
    await clone(remote, basename(local), dir);
  }

  await pull(local);
}

export function getPath(config: Config, name: string): string {
  let parts;
  if (name.length === 1) {
    parts = [1, name];
  } else if (name.length === 2) {
    parts = [2, name];
  } else if (name.length === 3) {
    parts = [3, name.substring(0, 1)];
  } else {
    parts = [name.substring(0, 2), name.substring(2, 4)];
  }

  return join(config.registry.local, ...parts, name);
}
