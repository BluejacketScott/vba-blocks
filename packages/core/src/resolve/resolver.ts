import { Config } from '../config';
import { Version, Dependency } from '../manifest';
import {
  isRegistryDependency,
  isPathDependency,
  isGitDependency
} from '../manifest/dependency';
import { Registration, registry, path, git } from '../manager';

export interface Resolution {
  name: string;
  range: string[];
  registered: Registration[];
}
export type ResolutionGraph = Map<string, Resolution>;

export default class Resolver {
  config: Config;
  graph: ResolutionGraph;
  loading: Map<string, Promise<Registration[]>>;

  constructor(config: Config) {
    this.graph = new Map();
  }

  async get(dependency: Dependency): Promise<Resolution> {
    const { name } = dependency;

    if (this.loading.has(name)) await this.loading.get(name);
    let resolution = this.graph.get(name);

    if (!resolution) {
      let loading;
      if (isRegistryDependency(dependency)) {
        loading = registry.resolve(this.config, dependency);
      } else if (isPathDependency(dependency)) {
        loading = path.resolve(this.config, dependency);
      } else {
        loading = git.resolve(this.config, dependency);
      }

      this.loading.set(name, loading);

      const registered = await loading;
      const resolution = {
        name,
        range: [],
        registered
      };

      this.loading.delete(name);
      this.graph.set(name, resolution);
    }

    let rangeInfo;
    if (isRegistryDependency(dependency)) {
      rangeInfo = dependency.version;
    } else if (isPathDependency(dependency)) {
      rangeInfo = dependency.path;
    } else {
      const { git, branch, tag, rev } = dependency;
      rangeInfo = { git, branch, tag, rev };
    }
    resolution.range.push(rangeInfo);

    return resolution;
  }

  getRegistration(id: string) {
    const [name, version] = id.split('@');
    const { registered } = this.graph.get(name);
    return registered.find(registration => registration.version === version);
  }

  [Symbol.iterator]() {
    return this.graph.entries();
  }
}
