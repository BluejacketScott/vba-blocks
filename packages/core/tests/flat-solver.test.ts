import { solve } from '../src/flat-solver';
import resolver from './helpers/resolver';

test('solves simple tree', async () => {
  const simple = {
    a: '^1.0.0'
  };
  const solution = await solve(simple, resolver);

  expect(solution.success).toEqual(true);
  expect(solution.result).toMatchSnapshot();
});
