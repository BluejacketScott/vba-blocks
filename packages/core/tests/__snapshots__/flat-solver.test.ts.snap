// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`solves simple tree 1`] = `
Array [
  "a@1.2.0",
  "d@2.0.0",
  "f@2.0.0",
  "g@1.0.0",
]
`;
